### PR TITLE
Update emscripten to 1.39.3

### LINKED ISSUE #6917: Consider updating emscripten 1.38.34 (resp. current version)
From 1.38.24:
``Perform JS static allocations at compile time (#7850)``
might be beneficial.

From 1.38.27:
``Remove deprecated Pointer_stringify (use UTF8ToString instead). See #8011``
might be a problem (we used that at some point, I'm not sure we still do).

### PR DESCRIPTION
**Pull request by ekpyron**
~~Depends on https://github.com/ethereum/solc-js/pull/418~~
Depends on https://github.com/ethereum/solidity/pull/7844
Closes #6917

### CODE REVIEW

**Comment by ekpyron** [comment]
Ok - the boost-related build failure I had locally apparently doesn't reproduce on CI, so it's only the solc-js change causing problems here... I'll try building to WASM to see what happens...

**Comment by axic** [comment]
This seems to reduce binary size:
- last build on develop (4 hrs ago): 14346146 Nov 28 11:08 soljson.js
- this build: 12295262 Nov 28 17:13 soljson.js

File: scripts/travis-emscripten/emscripten.jam [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# Boost.Build support for Emscipten.
```
Comment (axic) [inline scripts/travis-emscripten/emscripten.jam]: Please add the source link and license into this file or create a README.md explaining it.

File: test/externalTests/common.sh [hunk 2]
```diff
@@ -37,7 +37,7 @@ function setup_solcjs
 
     cd "$dir"
     printLog "Setting up solc-js..."
-    git clone --depth 1 -b v0.5.0 https://github.com/ethereum/solc-js.git solc
+    git clone --depth 1 -b emscripten-update https://github.com/ethereum/solc-js.git solc
```
Comment (axic) [inline test/externalTests/common.sh]: Need to be removed before merge.

File: .circleci/config.yml [hunk 3]
```diff
@@ -476,14 +476,14 @@ jobs:
...
-      - image: trzeci/emscripten:sdk-tag-1.38.22-64bit
+      - image: trzeci/emscripten:sdk-tag-1.39.3-64bit
     environment:
       TERM: xterm
     steps:
       - checkout
       - restore_cache:
           name: Restore Boost build
-          key: &boost-cache-key emscripten-boost-{{ checksum "scripts/travis-emscripten/install_deps.sh" }}{{ checksum "scripts/build_emscripten.sh" }}{{ checksum "scripts/travis-emscripten/build_emscripten.sh" }}
+          key: &boost-cache-key emscripten-boost-manual-cache-invalidation-{{ checksum "scripts/travis-emscripten/install_deps.sh" }}{{ checksum "scripts/build_emscripten.sh" }}{{ checksum "scripts/travis-emscripten/build_emscripten.sh" }}
```
Comment (axic) [inline .circleci/config.yml]: Needs to be removed before merge.

**Review by axic — CHANGES_REQUESTED** [review]
Please add a changelog entry.

**Comment by axic** [comment]
Marked it as 0.5.14, because it would be nice to get it in.

File: cmake/EthCompilerSettings.cmake [hunk 4]
```diff
@@ -86,7 +86,7 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 		elseif(EMSCRIPTEN)
 			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --memory-init-file 0")
 			# Leave only exported symbols as public and aggressively remove others
-			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fdata-sections -ffunction-sections -Wl,--gc-sections -fvisibility=hidden")
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake:89]: Did you check if there is a new option for gc-sections or LTO?

File: cmake/EthCompilerSettings.cmake [hunk 5]
```diff
@@ -110,9 +110,12 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s STRICT=1")
 			# Export the Emscripten-generated auxiliary methods which are needed by solc-js.
 			# Which methods of libsolc itself are exported is specified in libsolc/CMakeLists.txt.
-			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s EXTRA_EXPORTED_RUNTIME_METHODS=['cwrap','addFunction','removeFunction','Pointer_stringify','lengthBytesUTF8','_malloc','stringToUTF8','setValue']")
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s EXTRA_EXPORTED_RUNTIME_METHODS=['cwrap','addFunction','removeFunction','UTF8ToString','lengthBytesUTF8','_malloc','stringToUTF8','setValue']")
 			# Do not build as a WebAssembly target - we need an asm.js output.
 			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s WASM=0")
+
+			# Disable warnings about not being pure asm.js due to memory growth.
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-almost-asm")
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake]: This is actually not strongly related to the update, but is a prerequisite for the update.

File: cmake/EthCompilerSettings.cmake (hunk 4 above)
Comment (ekpyron) [reply cmake/EthCompilerSettings.cmake:89]: At least ``--gc-sections`` was/is apparently done by default.

**Review by axic — DISMISSED** [review]

**Comment by leonardoalt** [comment]
@axic not sure about the Travis test failure.

File: scripts/travis-emscripten/build_emscripten.sh [hunk 6]
```diff
@@ -55,11 +55,11 @@ fi
 WORKSPACE=/root/project
 
 # Increase nodejs stack size
-if ! [ -e /emsdk_portable/node/bin/node_orig ]
+if ! [ -e /emsdk_portable/node/current/bin/node_orig ]
 then
-  mv /emsdk_portable/node/bin/node /emsdk_portable/node/bin/node_orig
-  echo -e '#!/bin/sh\nexec /emsdk_portable/node/bin/node_orig --stack-size=8192 $@' > /emsdk_portable/node/bin/node
-  chmod 755 /emsdk_portable/node/bin/node
+  mv /emsdk_portable/node/current/bin/node /emsdk_portable/node/current/bin/node_orig
```
Comment (axic) [inline scripts/travis-emscripten/build_emscripten.sh:60]: @ekpyron travis is failing:
```
mv: cannot stat '/emsdk_portable/node/current/bin/node': No such file or directory

The command "test $SOLC_EMSCRIPTEN != On || (scripts/build_emscripten.sh)" failed and exited with 1 during .
```

Comment (axic) [reply scripts/travis-emscripten/build_emscripten.sh:60]: Ah, @ekpyron you forgot to update the docker image in travis.

**Comment by axic** [comment]
> not sure about the Travis test failure.

The docker image for travis wasn't updated.

**Comment by ekpyron** [comment]
Yeah, first I didn't care about travis, because it's somewhat in limbo anyways, and then I forgot about it, I'll look into it!

**Comment by axic** [comment]
> Yeah, first I didn't care about travis, because it's somewhat in limbo anyways, and then I forgot about it, I'll look into it!

I think I've fixed all of the issues.

**Comment by ekpyron** [comment]
(In case @axic's change didn't already fix it)

**Comment by ekpyron** [comment]
Damn, one shouldn't use github on the phone...

**Comment by axic** [comment]
> Damn, one shouldn't use github on the phone...

CI was almost finished. And it restarts with reopening. Just saying.

**Comment by ekpyron** [comment]
Yeah, sorry about that...

**Review by axic — APPROVED** [review]

**Pull request merged**